File: articles/image-captcha-solving-using-api/node/solve.js
/**
 * Solve Image CAPTCHA using the CaptchaAI API.
 *
 * Usage:
 *   npm install
 *   cp ../.env.example ../.env   // then edit with your credentials
 *   node solve.js
 */

require("dotenv").config({ path: require("path").join(__dirname, "..", ".env") });
const axios = require("axios");
const fs = require("fs");
const path = require("path");

const SUBMIT_URL = "https://ocr.captchaai.com/in.php";
const RESULT_URL = "https://ocr.captchaai.com/res.php";

const API_KEY = process.env.CAPTCHAAI_API_KEY || "";
const CAPTCHA_IMAGE_FILE = process.env.CAPTCHA_IMAGE_FILE || "captcha.png";
const POLL_INTERVAL = parseInt(process.env.POLL_INTERVAL || "5", 10);
const MAX_TIMEOUT = parseInt(process.env.MAX_TIMEOUT || "120", 10);

const AUTH_ERRORS = new Set(["ERROR_WRONG_USER_KEY", "ERROR_KEY_DOES_NOT_EXIST", "IP_BANNED"]);
const BALANCE_ERRORS = new Set(["ERROR_ZERO_BALANCE"]);
const INPUT_ERRORS = new Set([
  "ERROR_PAGEURL", "ERROR_WRONG_GOOGLEKEY", "ERROR_BAD_PARAMETERS",
  "ERROR_BAD_TOKEN_OR_PAGEURL", "ERROR_WRONG_FILE_EXTENSION",
  "ERROR_TOO_BIG_CAPTCHA_FILESIZE", "ERROR_EMPTY_CAPTCHA",
]);
const TRANSIENT_ERRORS = new Set(["ERROR_SERVER_ERROR", "ERROR_INTERNAL_SERVER_ERROR"]);
const SOLVE_ERRORS = new Set(["ERROR_CAPTCHA_UNSOLVABLE"]);
const PROXY_ERRORS = new Set(["ERROR_BAD_PROXY", "ERROR_PROXY_CONNECTION_FAILED"]);

function sleep(seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

function validateConfig() {
  if (!API_KEY || API_KEY === "YOUR_API_KEY") {
    console.log("[!] ERROR: CAPTCHAAI_API_KEY is not set.");
    console.log("    Copy .env.example to .env and add your real API key.");
    process.exit(1);
  }
  if (!CAPTCHA_IMAGE_FILE) {
    console.log("[!] ERROR: CAPTCHA_IMAGE_FILE is not set.");
    console.log("    Set the path to the CAPTCHA image file in .env.");
    process.exit(1);
  }
  const imgPath = resolveImagePath();
  if (!fs.existsSync(imgPath)) {
    console.log(`[!] ERROR: Image file not found: ${CAPTCHA_IMAGE_FILE}`);
    console.log("    Make sure the file exists and the path is correct.");
    process.exit(1);
  }
}

function resolveImagePath() {
  const parentPath = path.join(__dirname, "..", CAPTCHA_IMAGE_FILE);
  if (fs.existsSync(parentPath)) return parentPath;
  return CAPTCHA_IMAGE_FILE;
}

function loadImageBase64() {
  const imgPath = resolveImagePath();
  console.log(`[*] Loading image from ${CAPTCHA_IMAGE_FILE}...`);
  const data = fs.readFileSync(imgPath);
  return data.toString("base64");
}

async function submitTask(imageB64) {
  console.log("[*] Submitting Image CAPTCHA task...");

  const params = new URLSearchParams();
  params.append("key", API_KEY);
  params.append("method", "base64");
  params.append("body", imageB64);
  params.append("json", "1");

  let resp;
  try {
    resp = await axios.post(SUBMIT_URL, params.toString(), {
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      timeout: 30000,
    });
  } catch (err) {
    console.log(`[!] Network error during submission: ${err.message}`);
    process.exit(1);
  }

  const data = resp.data;
  if (data.status !== 1) {
    const error = data.request || "UNKNOWN_ERROR";
    if (AUTH_ERRORS.has(error)) {
      console.log(`[!] Authentication error: ${error}`);
      console.log("    Check your API key at https://captchaai.com/dashboard");
    } else if (BALANCE_ERRORS.has(error)) {
      console.log(`[!] Balance error: ${error}`);
      console.log("    Top up your account at https://captchaai.com");
    } else if (INPUT_ERRORS.has(error)) {
      console.log(`[!] Input error: ${error}`);
      console.log("    Verify your image file is valid and not too large.");
    } else if (PROXY_ERRORS.has(error)) {
      console.log(`[!] Proxy error: ${error}`);
      console.log("    Check your proxy configuration or try a different proxy.");
    } else {
      console.log(`[!] Submission failed: ${error}`);
    }
    process.exit(1);
  }

  const taskId = data.request;
  console.log(`[+] Task submitted. ID: ${taskId}`);
  return taskId;
}

async function pollResult(taskId) {
  console.log("[*] Waiting 10s before first poll...");
  await sleep(10);

  const params = { key: API_KEY, action: "get", id: taskId, json: "1" };
  let elapsed = 10;
  let attempt = 0;
  let backoff = POLL_INTERVAL;

  while (elapsed < MAX_TIMEOUT) {
    attempt++;
    console.log(`[*] Polling for result (attempt ${attempt})...`);

    let resp;
    try {
      resp = await axios.get(RESULT_URL, { params, timeout: 30000 });
    } catch (err) {
      console.log(`[!] Network error during polling: ${err.message}`);
      await sleep(backoff);
      elapsed += backoff;
      backoff = Math.min(backoff * 2, 30);
      continue;
    }

    const data = resp.data;

    if (data.status === 1) {
      return data.request;
    }

    const error = data.request || "";
    if (error === "CAPCHA_NOT_READY") {
      console.log(`[*] Not ready yet, waiting ${POLL_INTERVAL}s...`);
      await sleep(POLL_INTERVAL);
      elapsed += POLL_INTERVAL;
      backoff = POLL_INTERVAL;
      continue;
    }

    if (TRANSIENT_ERRORS.has(error)) {
      console.log(`[!] Transient error: ${error}, retrying in ${backoff}s...`);
      await sleep(backoff);
      elapsed += backoff;
      backoff = Math.min(backoff * 2, 30);
      continue;
    }

    if (SOLVE_ERRORS.has(error)) {
      console.log(`[!] Solve error: ${error}`);
      console.log("    The CAPTCHA could not be solved. Verify the image and retry.");
      process.exit(1);
    }

    console.log(`[!] Unexpected error: ${error}`);
    process.exit(1);
  }

  console.log(`[!] Timeout: no solution received within ${MAX_TIMEOUT} seconds.`);
  process.exit(1);
}

async function main() {
  validateConfig();
  const imageB64 = loadImageBase64();
  const taskId = await submitTask(imageB64);
  const text = await pollResult(taskId);
  console.log(`[+] Solved! Text: ${text}`);
  console.log();
  console.log("Next step: enter this text into the CAPTCHA input field");
  console.log("on the target page and submit the form.");
}

main();

File: articles/image-captcha-solving-using-api/python/solve.py
"""Solve Image CAPTCHA using the CaptchaAI API.

Usage:
    pip install -r requirements.txt
    cp ../.env.example ../.env   # then edit with your credentials
    python solve.py
"""

import os
import sys
import time
import base64

import requests
from dotenv import load_dotenv

# Load environment variables from the parent .env file
load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), "..", ".env"))

SUBMIT_URL = "https://ocr.captchaai.com/in.php"
RESULT_URL = "https://ocr.captchaai.com/res.php"

API_KEY = os.getenv("CAPTCHAAI_API_KEY", "")
CAPTCHA_IMAGE_FILE = os.getenv("CAPTCHA_IMAGE_FILE", "captcha.png")
POLL_INTERVAL = int(os.getenv("POLL_INTERVAL", "5"))
MAX_TIMEOUT = int(os.getenv("MAX_TIMEOUT", "120"))

# --- Error categories ---
AUTH_ERRORS = {"ERROR_WRONG_USER_KEY", "ERROR_KEY_DOES_NOT_EXIST", "IP_BANNED"}
BALANCE_ERRORS = {"ERROR_ZERO_BALANCE"}
INPUT_ERRORS = {
    "ERROR_PAGEURL", "ERROR_WRONG_GOOGLEKEY", "ERROR_BAD_PARAMETERS",
    "ERROR_BAD_TOKEN_OR_PAGEURL", "ERROR_WRONG_FILE_EXTENSION",
    "ERROR_TOO_BIG_CAPTCHA_FILESIZE", "ERROR_EMPTY_CAPTCHA",
}
TRANSIENT_ERRORS = {"ERROR_SERVER_ERROR", "ERROR_INTERNAL_SERVER_ERROR"}
SOLVE_ERRORS = {"ERROR_CAPTCHA_UNSOLVABLE"}
PROXY_ERRORS = {"ERROR_BAD_PROXY", "ERROR_PROXY_CONNECTION_FAILED"}


def validate_config() -> None:
    """Check that required environment variables are set."""
    if not API_KEY or API_KEY == "YOUR_API_KEY":
        print("[!] ERROR: CAPTCHAAI_API_KEY is not set.")
        print("    Copy .env.example to .env and add your real API key.")
        sys.exit(1)
    if not CAPTCHA_IMAGE_FILE:
        print("[!] ERROR: CAPTCHA_IMAGE_FILE is not set.")
        print("    Set the path to the CAPTCHA image file in .env.")
        sys.exit(1)
    image_path = os.path.join(os.path.dirname(__file__), "..", CAPTCHA_IMAGE_FILE)
    if not os.path.isfile(image_path) and not os.path.isfile(CAPTCHA_IMAGE_FILE):
        print(f"[!] ERROR: Image file not found: {CAPTCHA_IMAGE_FILE}")
        print("    Make sure the file exists and the path is correct.")
        sys.exit(1)


def load_image_base64() -> str:
    """Load image file and return base64-encoded string."""
    image_path = os.path.join(os.path.dirname(__file__), "..", CAPTCHA_IMAGE_FILE)
    if not os.path.isfile(image_path):
        image_path = CAPTCHA_IMAGE_FILE
    print(f"[*] Loading image from {CAPTCHA_IMAGE_FILE}...")
    with open(image_path, "rb") as f:
        return base64.b64encode(f.read()).decode("ascii")


def submit_task(image_b64: str) -> str:
    """Submit the CAPTCHA task via POST and return the task ID."""
    print("[*] Submitting Image CAPTCHA task...")
    payload = {
        "key": API_KEY,
        "method": "base64",
        "body": image_b64,
        "json": "1",
    }

    try:
        resp = requests.post(SUBMIT_URL, data=payload, timeout=30)
        resp.raise_for_status()
    except requests.RequestException as exc:
        print(f"[!] Network error during submission: {exc}")
        sys.exit(1)

    data = resp.json()

    if data.get("status") != 1:
        error = data.get("request", "UNKNOWN_ERROR")
        if error in AUTH_ERRORS:
            print(f"[!] Authentication error: {error}")
            print("    Check your API key at https://captchaai.com/dashboard")
        elif error in BALANCE_ERRORS:
            print(f"[!] Balance error: {error}")
            print("    Top up your account at https://captchaai.com")
        elif error in INPUT_ERRORS:
            print(f"[!] Input error: {error}")
            print("    Verify your image file is valid and not too large.")
        elif error in PROXY_ERRORS:
            print(f"[!] Proxy error: {error}")
            print("    Check your proxy configuration or try a different proxy.")
        else:
            print(f"[!] Submission failed: {error}")
        sys.exit(1)

    task_id = data["request"]
    print(f"[+] Task submitted. ID: {task_id}")
    return task_id


def poll_result(task_id: str) -> str:
    """Poll for the CAPTCHA solution with timeout and retry logic."""
    print("[*] Waiting 10s before first poll...")
    time.sleep(10)

    params = {
        "key": API_KEY,
        "action": "get",
        "id": task_id,
        "json": "1",
    }

    elapsed = 10
    attempt = 0
    backoff = POLL_INTERVAL

    while elapsed < MAX_TIMEOUT:
        attempt += 1
        print(f"[*] Polling for result (attempt {attempt})...")

        try:
            resp = requests.get(RESULT_URL, params=params, timeout=30)
            resp.raise_for_status()
        except requests.RequestException as exc:
            print(f"[!] Network error during polling: {exc}")
            time.sleep(backoff)
            elapsed += backoff
            backoff = min(backoff * 2, 30)
            continue

        data = resp.json()

        if data.get("status") == 1:
            return data["request"]

        error = data.get("request", "")
        if error == "CAPCHA_NOT_READY":
            print(f"[*] Not ready yet, waiting {POLL_INTERVAL}s...")
            time.sleep(POLL_INTERVAL)
            elapsed += POLL_INTERVAL
            backoff = POLL_INTERVAL
            continue

        if error in TRANSIENT_ERRORS:
            print(f"[!] Transient error: {error}, retrying in {backoff}s...")
            time.sleep(backoff)
            elapsed += backoff
            backoff = min(backoff * 2, 30)
            continue

        if error in SOLVE_ERRORS:
            print(f"[!] Solve error: {error}")
            print("    The CAPTCHA could not be solved. Verify the image and retry.")
            sys.exit(1)

        print(f"[!] Unexpected error: {error}")
        sys.exit(1)

    print(f"[!] Timeout: no solution received within {MAX_TIMEOUT} seconds.")
    sys.exit(1)


def main() -> None:
    validate_config()
    image_b64 = load_image_base64()
    task_id = submit_task(image_b64)
    text = poll_result(task_id)
    print(f"[+] Solved! Text: {text}")
    print()
    print("Next step: enter this text into the CAPTCHA input field")
    print("on the target page and submit the form.")


if __name__ == "__main__":
    main()
